resources/forum_metadata.py

```diff
@@ -0,0 +1,163 @@
+import aiohttp
+import discord
+from bs4 import BeautifulSoup
+from redbot.core import Config
+from redbot.core.commands import Context
+
+from .helper_functions import fetch
+from .user_profile import UserProfile
+
+
+class Metadata:
+    def __init__(self):
+        """This class is used to save character profiles so that they can be accessed without scraping the forumn.
+        This also makes it possible to get any character with posts by their display name.
+        """
+        self.config: Config = Config.get_conf(self, identifier=867530999999)
+        default_data = {
+            "character_profiles": {"0": None},
+            "unclaimed_characers": {"0": ""},
+            "no_posts": [],
+        }
+        self.config.init_custom("metadata", 1)
+        self.config.register_custom("metadata", **default_data)
+        self.profiles = UserProfile()
+
+    def get_profile(self, ctx, num):
+        """Gets a player profile from config as a discord embed object"""
+        profile_dict = self.config.custom("metadata", ctx.guild.id).character_profiles()
+        return discord.Embed.from_dict(profile_dict[num])
+
+    async def _fetch_character_profile(self, ctx, num, discord_name="Unkown"):
+        """takes an id and scrapes the forum for that character's profile.
+
+        Args:
+            num (int): The id of the character to search for.
+
+        Returns:
+            [str]: The name of the character, 'no posts' if there are no posts, or 'no char' if no character is found.
+        """
+        from .character import Character
+
+        discord_name = await self.get_discord_id_by_character_id(ctx, num)
+        recent_url = "http://themistborneisles.boards.net/user/" + str(num) + "/recent"
+        async with aiohttp.ClientSession() as session:
+            html = await fetch(session, recent_url)
+            soup_object = BeautifulSoup(html, "html.parser")
+        _profile = soup_object.find(class_="mini-profile")
+        if _profile:
+            return Character(_profile, soup_object, num, discord_name)
+
+        if "No posts were found." in str(soup_object):
+            return "no posts"
+        elif "The user you are trying to access could not be found." in str(
+            soup_object
+        ):
+            return "no char"
+
+    async def get_character(self, ctx, character):
+        """Gets a character profile by either dispay name or id.
+        If the character does not exist all characters will be updated.
```

Comment:
Why does it do this? Will this ever be called, like, if someone mistypes a name?

Reply:
This part is definitely complicated. Characters need to be stored locally in order for them to be searched by name.

Since you can't construct a link to a character profile if you only have a name and no way to get the number, the only thing you can do is check every single character. This takes a long time so the data has to be saved locally to speed tings up.

Whenever get_character is called, first the character name is searched locally. If it is not found then the forum is searched for new characters. If the bot comes across a character that is not saved then it saves it, compares the name to the name being searched, returns if the name is found, and continues if not.

So the first time this is run no profiles will be saved in metadata so it will take a long time. After that it will only ever update the most recent characters. 

Alternatively, if you search by character number only that page needs to be searched.

Reply:
If someone spells a name wrong it will try to update with that name, but wont find it

Reply:
I see. So when it does this, it starts at the end of the characters it knows about. We're not going to have situations where someone mistypes a name and it re-crawls all the characters,